independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/PublisherResponseHandler.java

```diff
@@ -46,32 +47,98 @@ public void setStreamingResponseCustomizers(List<StreamingResponseCustomizer> st
+        private volatile int state = 0;
+
         SseMultiSubscriber(ResteasyReactiveRequestContext requestContext, List<StreamingResponseCustomizer> staticCustomizers,
-                long demand) {
+                Publisher publisher, long demand) {
             super(requestContext, staticCustomizers, demand);
+            this.publisher = publisher;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            // we only request one item initially because we need to use that item to create the headers
+            // that will be sent in the first empty response
+            s.request(1);
+        }
+
+        @Override
+        public void onComplete() {
+            // make sure we don't trigger cancel with our onCloseHandler
+            weClosed = true;
+            // in this case the stream only had one item that we have yet to send (as we are waiting for the empty buffer to be sent)
+            if (state == 1) {
+                // do nothing as we still need to send the first item
+                // the connection will be closed by doSend when the item is sent
+            } else {
+                // no need to cancel on complete
+                // FIXME: are we interested in async completion?
+                doClose();
+            }
         }
 
         @Override
         public void onNext(Object item) {
-            OutboundSseEvent event;
-            if (item instanceof OutboundSseEvent) {
-                event = (OutboundSseEvent) item;
+            if (state == 0) {
+                state = 1;
+                SseUtil.setHeaders(requestContext, requestContext.serverResponse(),
+                        determineCustomizers(publisher, true, staticCustomizers));
+
+                requestContext.serverResponse().write(EMPTY_BUFFER, new Consumer<>() {
+                    @Override
+                    public void accept(Throwable throwable) {
+                        state = 2;
+                        if (throwable == null) {
+                            // now we can actually send the first item
+                            doSend(item);
+                        } else {
+                            requestContext.resume(throwable);
+                        }
+                    }
+                });
             } else {
-                event = new OutboundSseEventImpl.BuilderImpl().data(item).build();
+                doSend(item);
             }
-            SseUtil.send(requestContext, event, staticCustomizers).whenComplete((v, t) -> {
+        }
+
+        private void doSend(Object item) {
+            SseUtil.send(requestContext, fromItem(item), staticCustomizers).whenComplete((v, t) -> {
                 if (t != null) {
                     // need to cancel because the exception didn't come from the Multi
                     subscription.cancel();
```

Comment:
If you cancel, onComplete and onNext should not be called anymore - so you may need a proper terminal state.